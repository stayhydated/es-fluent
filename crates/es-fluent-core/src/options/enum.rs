use std::str::FromStr as _;

use crate::strategy::DisplayStrategy;
use bon::Builder;
use darling::{FromDeriveInput, FromField, FromMeta, FromVariant};
use getset::Getters;
use strum::IntoEnumIterator as _;

/// Options for an enum field.
#[derive(Clone, Debug, FromField, Getters)]
#[darling(attributes(fluent))]
pub struct EnumFieldOpts {
    /// The identifier of the field.
    #[getset(get = "pub")]
    ident: Option<syn::Ident>,
    /// The type of the field.
    #[getset(get = "pub")]
    ty: syn::Type,
    /// Whether to skip this field.
    #[darling(default)]
    skip: Option<bool>,
    /// Whether this field is a choice.
    #[darling(default)]
    choice: Option<bool>,
}

impl EnumFieldOpts {
    /// Returns `true` if the field should be skipped.
    pub fn is_skipped(&self) -> bool {
        self.skip.unwrap_or(false)
    }
    /// Returns `true` if the field is a choice.
    pub fn is_choice(&self) -> bool {
        self.choice.unwrap_or(false)
    }
}

/// Options for an enum variant.
#[derive(Clone, Debug, FromVariant, Getters)]
#[darling(attributes(fluent))]
pub struct VariantOpts {
    /// The identifier of the variant.
    #[getset(get = "pub")]
    ident: syn::Ident,
    fields: darling::ast::Fields<EnumFieldOpts>,
    /// Whether to skip this variant.
    #[darling(default)]
    skip: Option<bool>,
}

impl VariantOpts {
    /// Returns `true` if the variant should be skipped.
    pub fn is_skipped(&self) -> bool {
        self.skip.unwrap_or(false)
    }
    /// Returns the style of the variant's fields.
    pub fn style(&self) -> darling::ast::Style {
        self.fields.style
    }
    /// Returns the fields of the variant that are not skipped.
    pub fn fields(&self) -> Vec<&EnumFieldOpts> {
        self.fields
            .iter()
            .filter(|field| !field.is_skipped())
            .collect()
    }
    /// Returns all fields of the variant.
    pub fn all_fields(&self) -> Vec<&EnumFieldOpts> {
        self.fields.iter().collect()
    }
}

/// Options for an enum.
#[derive(Clone, FromDeriveInput, Getters, Debug)]
#[darling(supports(enum_unit, enum_named, enum_tuple), attributes(fluent))]
#[getset(get = "pub")]
pub struct EnumOpts {
    /// The identifier of the enum.
    ident: syn::Ident,
    /// The generics of the enum.
    generics: syn::Generics,
    data: darling::ast::Data<VariantOpts, darling::util::Ignored>,
    #[darling(flatten)]
    attr_args: EnumFluentAttributeArgs,
}

impl EnumOpts {
    /// Returns the variants of the enum.
    pub fn variants(&self) -> Vec<&VariantOpts> {
        match &self.data {
            darling::ast::Data::Enum(variants) => variants.iter().collect(),
            _ => unreachable!("Unexpected data type for enum"),
        }
    }
}

/// Attribute arguments for an enum.
#[derive(Builder, Clone, Debug, Default, FromMeta, Getters)]
pub struct EnumFluentAttributeArgs {
    display: Option<String>,
    #[darling(default)]
    choice: Option<bool>,
    #[darling(default)]
    this: Option<bool>,
}

impl EnumFluentAttributeArgs {
    /// Returns the display strategy for the enum.
    pub fn display(&self) -> DisplayStrategy {
        if let Some(mode_str) = self.display.as_deref() {
            DisplayStrategy::from_str(mode_str).unwrap_or_else(|_| {
                let possible_values: Vec<&'static str> = DisplayStrategy::iter()
                    .map(|variant| variant.into())
                    .collect();

                panic!(
                    "Unexpected mode: '{}', expected one of {:?}",
                    mode_str, possible_values
                );
            })
        } else {
            DisplayStrategy::FluentDisplay
        }
    }
    /// Returns `true` if the enum is a choice.
    pub fn is_choice(&self) -> bool {
        self.choice.unwrap_or(false)
    }
    /// Returns `true` if the enum should be passed as `this`.
    pub fn is_this(&self) -> bool {
        self.this.unwrap_or(false)
    }
}

/// Options for an enum that can be used as a choice.
#[derive(Clone, Debug, FromDeriveInput, Getters)]
#[darling(supports(enum_unit), attributes(fluent_choice))]
#[getset(get = "pub")]
pub struct EnumChoiceOpts {
    /// The identifier of the enum.
    ident: syn::Ident,
    /// The generics of the enum.
    generics: syn::Generics,
    data: darling::ast::Data<darling::util::Ignored, darling::util::Ignored>,
    #[darling(flatten)]
    attr_args: EnumChoiceAttributeArgs,
}

/// Attribute arguments for an enum that can be used as a choice.
#[derive(Builder, Clone, Debug, Default, FromMeta, Getters)]
#[getset(get = "pub")]
pub struct EnumChoiceAttributeArgs {
    #[darling(default)]
    serialize_all: Option<String>,
}
