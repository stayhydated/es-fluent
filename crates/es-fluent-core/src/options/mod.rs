pub mod r#enum;
pub mod r#struct;

const DEFAULT_FL_PATH: &str = "crate::fl";
