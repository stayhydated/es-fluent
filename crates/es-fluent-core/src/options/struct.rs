use std::str::FromStr as _;

use bon::Builder;
use darling::{FromDeriveInput, FromField, FromMeta};
use getset::Getters;
use quote::format_ident;
use strum::IntoEnumIterator as _;

use crate::strategy::DisplayStrategy;

#[derive(Clone, Debug, FromField, Getters)]
#[darling(attributes(fluent))]
pub struct StructFieldOpts {
    #[getset(get = "pub")]
    ident: Option<syn::Ident>,
    #[getset(get = "pub")]
    ty: syn::Type,
    #[darling(default)]
    skip: Option<bool>,
    #[darling(default)]
    default: Option<bool>,
}

impl StructFieldOpts {
    pub fn is_skipped(&self) -> bool {
        self.skip.unwrap_or(false)
    }

    pub fn is_default(&self) -> bool {
        self.default.unwrap_or(false)
    }
}

#[derive(Clone, FromDeriveInput, Getters)]
#[darling(supports(struct_named), attributes(fluent))]
#[getset(get = "pub")]
pub struct StructOpts {
    ident: syn::Ident,
    #[darling(default)]
    derive: darling::util::PathList,
    generics: syn::Generics,
    data: darling::ast::Data<darling::util::Ignored, StructFieldOpts>,
    #[darling(flatten)]
    attr_args: StructFluentAttributeArgs,
}

impl StructOpts {
    const FTL_ENUM_IDENT: &str = "Ftl";

    pub fn ftl_enum_ident(&self) -> syn::Ident {
        format_ident!("{}{}", &self.ident, Self::FTL_ENUM_IDENT)
    }

    pub fn keyyed_idents(&self) -> Vec<syn::Ident> {
        self.attr_args
            .clone()
            .keys
            .map_or_else(Vec::new, |keys| {
                keys.into_iter()
                    .map(|key| format_ident!("{}", key.value()))
                    .collect()
            })
            .into_iter()
            .map(|key| format_ident!("{}{}{}", &self.ident, key, Self::FTL_ENUM_IDENT))
            .collect()
    }

    pub fn fields(&self) -> Vec<&StructFieldOpts> {
        match &self.data {
            darling::ast::Data::Struct(fields) => fields
                .fields
                .iter()
                .filter(|field| !field.is_skipped())
                .collect(),
            _ => vec![],
        }
    }
}

#[derive(Builder, Clone, Debug, Default, FromMeta, Getters)]
pub struct StructFluentAttributeArgs {
    #[darling(default)]
    fl: Option<syn::Path>, // this is probably useless, force user to provide a fl! macro?
    display: Option<String>,
    #[darling(default)]
    keys: Option<Vec<syn::LitStr>>,
}
impl StructFluentAttributeArgs {
    pub fn fl(&self) -> syn::Path {
        let default_fl_path: syn::Path = syn::parse_str(super::DEFAULT_FL_PATH).unwrap();
        self.fl.clone().unwrap_or(default_fl_path)
    }
    pub fn display(&self) -> DisplayStrategy {
        if let Some(mode_str) = self.display.as_deref() {
            DisplayStrategy::from_str(mode_str).unwrap_or_else(|_| {
                let possible_values: Vec<&'static str> = DisplayStrategy::iter()
                    .map(|variant| variant.into())
                    .collect();

                panic!(
                    "Unexpected mode: '{}', expected one of {:?}",
                    mode_str, possible_values
                );
            })
        } else {
            DisplayStrategy::StdDisplay
        }
    }
}
