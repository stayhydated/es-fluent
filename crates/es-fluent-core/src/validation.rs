//! This module provides functions for validating `es-fluent` attributes.

use crate::error::{ErrorExt as _, EsFluentCoreError, EsFluentCoreResult};
use crate::options::r#enum::EnumOpts;
use crate::options::r#struct::StructOpts;
use syn::{DataEnum, DataStruct};

/// Validates the `es-fluent` attributes on an enum.
pub fn validate_enum(_opts: &EnumOpts, _data: &DataEnum) -> EsFluentCoreResult<()> {
    Ok(())
}

/// Validates the `es-fluent` attributes on a struct.
pub fn validate_struct(opts: &StructOpts, _data: &DataStruct) -> EsFluentCoreResult<()> {
    validate_struct_defaults(opts)?;
    Ok(())
}

fn validate_struct_defaults(opts: &StructOpts) -> EsFluentCoreResult<()> {
    let default_fields: Vec<_> = opts
        .indexed_fields()
        .into_iter()
        .filter(|(_, field)| field.is_default())
        .collect();

    if default_fields.len() > 1 {
        let (first_index, first_field) = &default_fields[0];
        let (second_index, second_field) = &default_fields[1];

        let first_field_name = first_field.fluent_arg_name(*first_index);
        let second_field_name = second_field.fluent_arg_name(*second_index);
        let second_span = second_field.ident().as_ref().map(|ident| ident.span());

        return Err(EsFluentCoreError::FieldError {
            message: "Struct cannot have multiple fields marked `#[fluent(default)]`.".to_string(),
            field_name: Some(second_field_name),
            span: second_span,
        }
        .with_note(format!(
            "First `#[fluent(default)]` field found was `{}`.",
            first_field_name
        )));
    }
    Ok(())
}
