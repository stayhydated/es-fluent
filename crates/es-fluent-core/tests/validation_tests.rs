use darling::FromDeriveInput;
use es_fluent_core::options::r#struct::StructOpts;
use es_fluent_core::validation::validate_struct;
use syn::{DataStruct, DeriveInput, parse_quote};

#[test]
fn validate_struct_multiple_defaults_produces_expected_error_message() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluent)]
        pub struct TestStruct {
            #[fluent(default)]
            field1: String,
            #[fluent(default)]
            field2: i32,
        }
    };

    let opts = StructOpts::from_derive_input(&input).expect("StructOpts should parse");
    let data = DataStruct {
        struct_token: Default::default(),
        fields: syn::Fields::Named(parse_quote! { { field1: String, field2: i32 } }),
        semi_token: None,
    };

    let err = validate_struct(&opts, &data).expect_err("Expected validation error");

    insta::assert_ron_snapshot!(
        "validate_struct_multiple_defaults_produces_expected_error_message",
        err.to_string()
    );
}

#[test]
fn validate_struct_multiple_defaults_tuple_struct_produces_expected_error_message() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluent)]
        pub struct TestTupleStruct(
            #[fluent(default)]
            String,
            #[fluent(default)]
            i32,
        );
    };

    let opts = StructOpts::from_derive_input(&input).expect("StructOpts should parse");
    let data = match &input.data {
        syn::Data::Struct(data) => data,
        _ => unreachable!("input is a struct"),
    };

    let err = validate_struct(&opts, data).expect_err("Expected validation error");

    insta::assert_ron_snapshot!(
        "validate_struct_multiple_defaults_tuple_struct_produces_expected_error_message",
        err.to_string()
    );
}
