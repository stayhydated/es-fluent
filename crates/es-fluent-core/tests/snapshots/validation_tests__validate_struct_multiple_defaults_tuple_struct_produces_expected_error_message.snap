---
source: crates/es-fluent-core/tests/validation_tests.rs
expression: err.to_string()
---
"Field \'f1\' error: Struct cannot have multiple fields marked `#[fluent(default)]`.\nnote: First `#[fluent(default)]` field found was `f0`."
